packages/Kokkos/src/DTK_PointCloud_def.hpp

```diff
@@ -0,0 +1,235 @@
+/*!
+  \brief Constructor.
+*/
+template <class SC, class LO, class GO, class NO>
+PointCloud<SC, LO, GO, NO>::PointCloud(
+    const Teuchos::RCP<const Teuchos::Comm<int>> &comm,
+    const global_id_view global_ids, const coordinate_view coordinates )
+    : _comm( comm )
+    , _global_ids( global_ids )
+    , _coordinates( coordinates )
+{
+    DTK_REQUIRE( _global_ids.extent( 0 ) == _coordinates.extent( 0 ) );
+}
+
+//---------------------------------------------------------------------------//
+// Get the spatial dimension of the point cloud.
+template <class SC, class LO, class GO, class NO>
+size_t PointCloud<SC, LO, GO, NO>::spaceDim() const
+{
+    return _coordinates.extent( 1 );
+}
+
+//---------------------------------------------------------------------------//
+// Get the local number of points in the point cloud.
+template <class SC, class LO, class GO, class NO>
+size_t PointCloud<SC, LO, GO, NO>::numLocalPoints() const
+{
+    return _global_ids.extent( 0 );
+}
+
+//---------------------------------------------------------------------------//
+// Get the global number of points in the point cloud.
+template <class SC, class LO, class GO, class NO>
+size_t PointCloud<SC, LO, GO, NO>::numGlobalPoints() const
+{
+    size_t local_num = numLocalPoints();
+    size_t global_num = 0;
+    Teuchos::reduceAll( *_comm, Teuchos::REDUCE_SUM, local_num,
+                        Teuchos::ptrFromRef( global_num ) );
+    return global_num;
+}
+
+//---------------------------------------------------------------------------//
+// Get the local bounding box.
+template <class SC, class LO, class GO, class NO>
+Kokkos::Array<SC, 6> PointCloud<SC, LO, GO, NO>::localBoundingBox() const
+{
+    SC local_box[6];
+
+    LocalBoundingBoxFunctor<SC, LO, GO, NO> functor( _coordinates );
+
+    Kokkos::parallel_reduce(
+        "Point cloud local bounding box",
+        Kokkos::RangePolicy<execution_space>( 0, numLocalPoints() ), functor,
+        local_box );
+    Kokkos::fence();
+
+    Kokkos::Array<SC, 6> box_array;
+    for ( int i = 0; i < 6; ++i )
+        box_array[i] = local_box[i];
+    return box_array;
+}
+
+//---------------------------------------------------------------------------//
+// Get the global bounding box.
+template <class SC, class LO, class GO, class NO>
+Kokkos::Array<SC, 6> PointCloud<SC, LO, GO, NO>::globalBoundingBox() const
+{
+    Kokkos::Array<SC, 6> local_box = localBoundingBox();
+    Kokkos::Array<SC, 6> global_box;
+
+    Teuchos::reduceAll( *_comm, Teuchos::REDUCE_MIN, 3, &local_box[0],
+                        &global_box[0] );
+
+    Teuchos::reduceAll( *_comm, Teuchos::REDUCE_MAX, 3, &local_box[3],
+                        &global_box[3] );
+
+    return global_box;
+}
+
+//---------------------------------------------------------------------------//
+// Get the point global ids.
+template <class SC, class LO, class GO, class NO>
+const auto PointCloud<SC, LO, GO, NO>::globalIds() const -> global_id_view
```

Comment:
it lets one process the return type after the function definition. so you could use decltype based on input and template arguments or in this case use a type alias directly